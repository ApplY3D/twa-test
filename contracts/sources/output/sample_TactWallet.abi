{"name":"TactWallet","types":[{"name":"StateInit","header":null,"fields":[{"name":"code","type":{"kind":"simple","type":"cell","optional":false}},{"name":"data","type":{"kind":"simple","type":"cell","optional":false}}]},{"name":"Context","header":null,"fields":[{"name":"bounced","type":{"kind":"simple","type":"bool","optional":false}},{"name":"sender","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"raw","type":{"kind":"simple","type":"slice","optional":false}}]},{"name":"SendParameters","header":null,"fields":[{"name":"bounce","type":{"kind":"simple","type":"bool","optional":false}},{"name":"to","type":{"kind":"simple","type":"address","optional":false}},{"name":"value","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"mode","type":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"body","type":{"kind":"simple","type":"cell","optional":true}},{"name":"code","type":{"kind":"simple","type":"cell","optional":true}},{"name":"data","type":{"kind":"simple","type":"cell","optional":true}}]},{"name":"ChangeOwner","header":2174598809,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"ChangeOwnerOk","header":846932810,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"newOwner","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"Deploy","header":2490013878,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"DeployOk","header":2952335191,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}}]},{"name":"FactoryDeploy","header":1829761339,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"cashback","type":{"kind":"simple","type":"address","optional":false}}]},{"name":"Array","header":null,"fields":[{"name":"map","type":{"kind":"dict","key":"uint","keyFormat":16,"value":"int"}},{"name":"length","type":{"kind":"simple","type":"int","optional":false,"format":257}}]},{"name":"ExtMessage","header":3240438462,"fields":[{"name":"signature","type":{"kind":"simple","type":"fixed-bytes","optional":false,"format":64}},{"name":"seqno","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"valid_until","type":{"kind":"simple","type":"uint","optional":false,"format":32}},{"name":"message_parameters","type":{"kind":"simple","type":"SendParameters","optional":false}}]},{"name":"TokenTransfer","header":260734629,"fields":[{"name":"queryId","type":{"kind":"simple","type":"uint","optional":false,"format":64}},{"name":"amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"destination","type":{"kind":"simple","type":"address","optional":false}},{"name":"response_destination","type":{"kind":"simple","type":"address","optional":false}},{"name":"custom_payload","type":{"kind":"simple","type":"cell","optional":true}},{"name":"forward_ton_amount","type":{"kind":"simple","type":"uint","optional":false,"format":"coins"}},{"name":"forward_payload","type":{"kind":"simple","type":"slice","optional":false,"format":"remainder"}}]},{"name":"AddAddress","header":141837752,"fields":[{"name":"public_key","type":{"kind":"simple","type":"int","optional":false,"format":257}}]}],"receivers":[{"receiver":"internal","message":{"kind":"empty"}},{"receiver":"internal","message":{"kind":"text"}},{"receiver":"internal","message":{"kind":"any"}},{"receiver":"internal","message":{"kind":"typed","type":"AddAddress"}},{"receiver":"external","message":{"kind":"typed","type":"ExtMessage"}},{"receiver":"internal","message":{"kind":"typed","type":"Deploy"}},{"receiver":"internal","message":{"kind":"typed","type":"ChangeOwner"}}],"getters":[{"name":"seqno","arguments":[],"returnType":{"kind":"simple","type":"int","optional":false,"format":257}},{"name":"get_public_keys","arguments":[],"returnType":{"kind":"dict","key":"uint","keyFormat":16,"value":"int"}},{"name":"owner","arguments":[],"returnType":{"kind":"simple","type":"address","optional":false}}],"errors":{"2":{"message":"Stack undeflow"},"3":{"message":"Stack overflow"},"4":{"message":"Integer overflow"},"5":{"message":"Integer out of expected range"},"6":{"message":"Invalid opcode"},"7":{"message":"Type check error"},"8":{"message":"Cell overflow"},"9":{"message":"Cell underflow"},"10":{"message":"Dictionary error"},"13":{"message":"Out of gas error"},"32":{"message":"Method ID not found"},"34":{"message":"Action is invalid or not supported"},"37":{"message":"Not enough TON"},"38":{"message":"Not enough extra-currencies"},"128":{"message":"Null reference exception"},"129":{"message":"Invalid serialization prefix"},"130":{"message":"Invalid incoming message"},"131":{"message":"Constraints error"},"132":{"message":"Access denied"},"133":{"message":"Contract stopped"},"134":{"message":"Invalid argument"},"135":{"message":"Code of a contract was not found"},"136":{"message":"Invalid address"},"137":{"message":"Masterchain support is not enabled for this contract"},"6947":{"message":"No items in the array!"},"9577":{"message":"Invalid Signature"},"12033":{"message":"Invalid Time"},"17848":{"message":"No items in the array to delete!"},"19333":{"message":"Index of the item cannot be negative!"},"24006":{"message":"Index is out of array bounds!"},"30508":{"message":"Invalid Seqno"},"42469":{"message":"No space in the array left for new items!"}},"interfaces":["org.ton.introspection.v0","org.ton.abi.ipfs.v0","org.ton.deploy.lazy.v0","org.ton.chain.workchain.v0","org.ton.ownable.transferable.v2","org.ton.ownable"]}