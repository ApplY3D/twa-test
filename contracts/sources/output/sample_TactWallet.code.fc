#pragma version =0.4.3;
#pragma allow-post-modification;
#pragma compute-asm-ltr;

#include "sample_TactWallet.headers.fc";
#include "sample_TactWallet.stdlib.fc";
#include "sample_TactWallet.storage.fc";

;;
;; Contract TactWallet functions
;;

(slice, int, (cell, int), int) $TactWallet$_contract_init((cell, int) $publicKeys) impure inline_ref {
    var (($publicKeys'map, $publicKeys'length)) = $publicKeys;
    var (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)) = (null(), 0, null(), null());
    $self'owner = __tact_context_get_sender();
    ($self'publicKeys'map, $self'publicKeys'length) = ($publicKeys'map, $publicKeys'length);
    $self'deployTime = now();
    return ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime);
}

((cell, int), ()) $Array$_fun_append((cell, int) $self, int $item) impure inline_ref {
    var (($self'map, $self'length)) = $self;
    var (($self'map, $self'length)) = $self;
    throw_unless(42469, (($self'length + 1) <= 10));
    $self'map~__tact_dict_set_uint_int(16, $self'length, $item, 257);
    $self'length = ($self'length + 1);
    return (($self'map, $self'length), ());
}

int $Array$_fun_getIdx((cell, int) $self, int $idx) impure inline_ref {
    var (($self'map, $self'length)) = $self;
    var (($self'map, $self'length)) = $self;
    throw_unless(6947, ($self'length > 0));
    throw_unless(19333, ($idx >= 0));
    throw_unless(24006, ($idx < $self'length));
    return __tact_not_null(__tact_dict_get_uint_int($self'map, 16, $idx, 257));
}

((slice, int, (cell, int), int), int) $TactWallet$_fun_seqno((slice, int, (cell, int), int) $self) impure inline_ref {
    var (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), $self'seqno);
}

((slice, int, (cell, int), int), cell) $TactWallet$_fun_get_public_keys((slice, int, (cell, int), int) $self) impure inline_ref {
    var (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), $self'publicKeys'map);
}

((slice, int, (cell, int), int), ()) $TactWallet$_fun_requireOwner((slice, int, (cell, int), int) $self) impure inline_ref {
    var (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)) = $self;
    throw_unless(132, ( __tact_slice_eq_bits($self'owner, __tact_context_get_sender()) ));
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

((slice, int, (cell, int), int), slice) $TactWallet$_fun_owner((slice, int, (cell, int), int) $self) impure inline_ref {
    var (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), $self'owner);
}

;;
;; Receivers of a Contract TactWallet
;;

(((slice, int, (cell, int), int)), ()) %$TactWallet$_internal_empty((slice, int, (cell, int), int) $self) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

((slice, int, (cell, int), int), ()) $TactWallet$_internal_any_text((slice, int, (cell, int), int) $self, slice $str) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

((slice, int, (cell, int), int), ()) $TactWallet$_internal_any((slice, int, (cell, int), int) $self, slice $slice) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

(((slice, int, (cell, int), int)), ()) $TactWallet$_internal_binary_AddAddress((slice, int, (cell, int), int) $self, (int) $msg) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    var ($msg'public_key) = $msg;
    ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)~$TactWallet$_fun_requireOwner();
    ($self'publicKeys'map, $self'publicKeys'length)~$Array$_fun_append($msg'public_key);
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

(((slice, int, (cell, int), int)), ()) $TactWallet$_external_binary_ExtMessage((slice, int, (cell, int), int) $self, (slice, int, int, (int, slice, int, int, cell, cell, cell)) $msg) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    var ($msg'signature, $msg'seqno, $msg'valid_until, ($msg'message_parameters'bounce, $msg'message_parameters'to, $msg'message_parameters'value, $msg'message_parameters'mode, $msg'message_parameters'body, $msg'message_parameters'code, $msg'message_parameters'data)) = $msg;
    int $hash = cell_hash(end_cell(store_ref(store_uint(store_uint(begin_cell(), $msg'seqno, 32), $msg'valid_until, 32), $SendParameters$_store_cell(($msg'message_parameters'bounce, $msg'message_parameters'to, $msg'message_parameters'value, $msg'message_parameters'mode, $msg'message_parameters'body, $msg'message_parameters'code, $msg'message_parameters'data)))));
    int $isValid = false;
    int $index = 0;
    while (($index < $self'publicKeys'length)) {
        if (check_signature($hash, $msg'signature, $Array$_fun_getIdx(($self'publicKeys'map, $self'publicKeys'length), $index))) {
            $isValid = true;
            $index = $self'publicKeys'length;
        }
        $index = ($index + 1);
    }
    throw_unless(9577, $isValid);
    throw_unless(30508, ($msg'seqno == $self'seqno));
    throw_unless(12033, (now() <= $msg'valid_until));
    accept_message();
    $self'seqno = ($self'seqno + 1);
    $global_send(($msg'message_parameters'bounce, $msg'message_parameters'to, $msg'message_parameters'value, $msg'message_parameters'mode, $msg'message_parameters'body, $msg'message_parameters'code, $msg'message_parameters'data));
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

(((slice, int, (cell, int), int)), ()) $TactWallet$_internal_binary_Deploy((slice, int, (cell, int), int) $self, (int) $deploy) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    var ($deploy'queryId) = $deploy;
    ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)~$TactWallet$_fun_notify($DeployOk$_store_cell($DeployOk$_constructor_queryId($deploy'queryId)));
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

(((slice, int, (cell, int), int)), ()) $TactWallet$_internal_binary_ChangeOwner((slice, int, (cell, int), int) $self, (int, slice) $msg) impure inline {
    var ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime) = $self;
    var ($msg'queryId, $msg'newOwner) = $msg;
    ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)~$TactWallet$_fun_requireOwner();
    $self'owner = $msg'newOwner;
    ($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime)~$TactWallet$_fun_reply($ChangeOwnerOk$_store_cell($ChangeOwnerOk$_constructor_queryId_newOwner($msg'queryId, $msg'newOwner)));
    return (($self'owner, $self'seqno, ($self'publicKeys'map, $self'publicKeys'length), $self'deployTime), ());
}

;;
;; Get methods of a Contract TactWallet
;;

_ %seqno() method_id(85143) {
    var self = $TactWallet$_contract_load();
    var res = self~$TactWallet$_fun_seqno();
    return res;
}

_ %get_public_keys() method_id(120004) {
    var self = $TactWallet$_contract_load();
    var res = self~$TactWallet$_fun_get_public_keys();
    return res;
}

_ %owner() method_id(83229) {
    var self = $TactWallet$_contract_load();
    var res = self~$TactWallet$_fun_owner();
    return res;
}

_ supported_interfaces() method_id {
    return (
        "org.ton.introspection.v0"H >> 128,
        "org.ton.abi.ipfs.v0"H >> 128,
        "org.ton.deploy.lazy.v0"H >> 128,
        "org.ton.chain.workchain.v0"H >> 128,
        "org.ton.ownable.transferable.v2"H >> 128,
        "org.ton.ownable"H >> 128
    );
}

_ get_abi_ipfs() method_id {
    return "ipfs://QmdTGzNJs9tXntGPPRYgpcSzHcpQFgNcKmd8uafPSWoRAs";
}

_ lazy_deployment_completed() method_id {
    return get_data().begin_parse().load_int(1);
}

;;
;; Routing of a Contract TactWallet
;;

((slice, int, (cell, int), int), int) $TactWallet$_contract_router_internal((slice, int, (cell, int), int) self, int msg_bounced, slice in_msg) impure inline_ref {
    ;; Handle bounced messages
    if (msg_bounced) {
        return (self, true);
    }
    
    ;; Parse incoming message
    int op = 0;
    if (slice_bits(in_msg) >= 32) {
        op = in_msg.preload_uint(32);
    }
    
    
    ;; Receive empty message
    if ((op == 0) & (slice_bits(in_msg) <= 32)) {
        self~%$TactWallet$_internal_empty();
        return (self, true);
    }
    
    ;; Receive AddAddress message
    if (op == 141837752) {
        var msg = in_msg~$AddAddress$_load();
        self~$TactWallet$_internal_binary_AddAddress(msg);
        return (self, true);
    }
    
    ;; Receive Deploy message
    if (op == 2490013878) {
        var msg = in_msg~$Deploy$_load();
        self~$TactWallet$_internal_binary_Deploy(msg);
        return (self, true);
    }
    
    ;; Receive ChangeOwner message
    if (op == 2174598809) {
        var msg = in_msg~$ChangeOwner$_load();
        self~$TactWallet$_internal_binary_ChangeOwner(msg);
        return (self, true);
    }
    
    ;; Text Receivers
    if (op == 0) {
        if (slice_bits(in_msg) >= 32) {
            self~$TactWallet$_internal_any_text(in_msg.skip_bits(32));
            return (self, true);
        }
    }
    
    ;; Receiver fallback
    self~$TactWallet$_internal_any(in_msg);
    return (self, true);
}

((slice, int, (cell, int), int), int) $TactWallet$_contract_router_external((slice, int, (cell, int), int) self, slice in_msg) impure inline_ref {
    
    ;; Parse incoming message
    int op = 0;
    if (slice_bits(in_msg) >= 32) {
        op = in_msg.preload_uint(32);
    }
    
    
    ;; Receive ExtMessage message
    if (op == 3240438462) {
        var msg = in_msg~$ExtMessage$_load();
        self~$TactWallet$_external_binary_ExtMessage(msg);
        return (self, true);
    }
    
    return (self, false);
}

() recv_internal(int msg_value, cell in_msg_cell, slice in_msg) impure {
    
    ;; Context
    var cs = in_msg_cell.begin_parse();
    var msg_flags = cs~load_uint(4);
    var msg_bounced = -(msg_flags & 1);
    slice msg_sender_addr = __tact_verify_address(cs~load_msg_addr());
    __tact_context = (msg_bounced, msg_sender_addr, msg_value, cs);
    __tact_context_sender = msg_sender_addr;
    
    ;; Load contract data
    var self = $TactWallet$_contract_load();
    
    ;; Handle operation
    int handled = self~$TactWallet$_contract_router_internal(msg_bounced, in_msg);
    
    ;; Throw if not handled
    throw_unless(130, handled);
    
    ;; Persist state
    $TactWallet$_contract_store(self);
}

() recv_external(slice in_msg) impure {
    ;; Load contract data
    var self = $TactWallet$_contract_load();
    
    ;; Handle operation
    int handled = self~$TactWallet$_contract_router_external(in_msg);
    
    ;; Throw if not handled
    throw_unless(handled, 130);
    
    ;; Persist state
    $TactWallet$_contract_store(self);
}
