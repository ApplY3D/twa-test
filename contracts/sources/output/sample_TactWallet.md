# TACT Compilation Report
Contract: TactWallet
BOC Size: 1536 bytes

# Types
Total Types: 12

## StateInit
TLB: `_ code:^cell data:^cell = StateInit`
Signature: `StateInit{code:^cell,data:^cell}`

## Context
TLB: `_ bounced:bool sender:address value:int257 raw:^slice = Context`
Signature: `Context{bounced:bool,sender:address,value:int257,raw:^slice}`

## SendParameters
TLB: `_ bounce:bool to:address value:int257 mode:int257 body:Maybe ^cell code:Maybe ^cell data:Maybe ^cell = SendParameters`
Signature: `SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}`

## ChangeOwner
TLB: `change_owner#819dbe99 queryId:uint64 newOwner:address = ChangeOwner`
Signature: `ChangeOwner{queryId:uint64,newOwner:address}`

## ChangeOwnerOk
TLB: `change_owner_ok#327b2b4a queryId:uint64 newOwner:address = ChangeOwnerOk`
Signature: `ChangeOwnerOk{queryId:uint64,newOwner:address}`

## Deploy
TLB: `deploy#946a98b6 queryId:uint64 = Deploy`
Signature: `Deploy{queryId:uint64}`

## DeployOk
TLB: `deploy_ok#aff90f57 queryId:uint64 = DeployOk`
Signature: `DeployOk{queryId:uint64}`

## FactoryDeploy
TLB: `factory_deploy#6d0ff13b queryId:uint64 cashback:address = FactoryDeploy`
Signature: `FactoryDeploy{queryId:uint64,cashback:address}`

## Array
TLB: `_ map:dict<uint16, int> length:int257 = Array`
Signature: `Array{map:dict<uint16, int>,length:int257}`

## ExtMessage
TLB: `ext_message#c1252abe signature:fixed_bytes64 seqno:uint32 valid_until:uint32 message_parameters:SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell} = ExtMessage`
Signature: `ExtMessage{signature:fixed_bytes64,seqno:uint32,valid_until:uint32,message_parameters:SendParameters{bounce:bool,to:address,value:int257,mode:int257,body:Maybe ^cell,code:Maybe ^cell,data:Maybe ^cell}}`

## TokenTransfer
TLB: `token_transfer#0f8a7ea5 queryId:uint64 amount:coins destination:address response_destination:address custom_payload:Maybe ^cell forward_ton_amount:coins forward_payload:remainder<slice> = TokenTransfer`
Signature: `TokenTransfer{queryId:uint64,amount:coins,destination:address,response_destination:address,custom_payload:Maybe ^cell,forward_ton_amount:coins,forward_payload:remainder<slice>}`

## AddAddress
TLB: `add_address#087445b8 public_key:int257 = AddAddress`
Signature: `AddAddress{public_key:int257}`

# Get Methods
Total Get Methods: 3

## seqno

## get_public_keys

## owner

# Error Codes
2: Stack undeflow
3: Stack overflow
4: Integer overflow
5: Integer out of expected range
6: Invalid opcode
7: Type check error
8: Cell overflow
9: Cell underflow
10: Dictionary error
13: Out of gas error
32: Method ID not found
34: Action is invalid or not supported
37: Not enough TON
38: Not enough extra-currencies
128: Null reference exception
129: Invalid serialization prefix
130: Invalid incoming message
131: Constraints error
132: Access denied
133: Contract stopped
134: Invalid argument
135: Code of a contract was not found
136: Invalid address
137: Masterchain support is not enabled for this contract
6947: No items in the array!
9577: Invalid Signature
12033: Invalid Time
17848: No items in the array to delete!
19333: Index of the item cannot be negative!
24006: Index is out of array bounds!
30508: Invalid Seqno
42469: No space in the array left for new items!