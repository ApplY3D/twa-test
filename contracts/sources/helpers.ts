import { Address, Cell, beginCell } from "ton";
import { sign } from "ton-crypto";
import { SendParameters, storeSendParameters, TokenTransfer, storeTokenTransfer } from "./output/sample_TactWallet";

export function fill_send_parameters(
    to: Address,
    value: bigint,
    body: Cell,
    mode: bigint = 1n,
    bounce: boolean = true,
    code: Cell | null = null,
    data: Cell | null = null
): SendParameters {
    return { $$type: "SendParameters", to, value, body, mode, bounce, code, data };
}

export function jetton_send_parameters(data: {
    queryId: bigint;
    amount: bigint;
    destination: Address;
    response_destination: Address;
    custom_payload: Cell | null;
    forward_ton_amount: bigint;
    forward_payload: Cell;
}): TokenTransfer {
    return { $$type: "TokenTransfer", ...data };
}

export async function send_ext_message(
    wallet: any,
    secretKey: Buffer,
    valid_until: bigint,
    params: SendParameters | TokenTransfer
) {
    let parameters_b = beginCell();
    if (params.$$type === "TokenTransfer") {
        storeTokenTransfer(params)(parameters_b);
    } else {
        storeSendParameters(params)(parameters_b);
    }
    let seqno = await wallet.getSeqno();
    let cell = beginCell().storeUint(seqno, 32).storeUint(valid_until, 32).storeRef(parameters_b.endCell()).endCell();
    const hash = cell.hash();
    wallet.sendExternal({
        $$type: "ExtMessage",
        signature: sign(hash, secretKey),
        seqno,
        valid_until,
        message_parameters: params,
    });
}
