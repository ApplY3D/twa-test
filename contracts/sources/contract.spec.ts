import { beginCell, toNano } from "ton";
import { mnemonicNew, mnemonicToWalletKey } from "ton-crypto";
import { ContractSystem } from "@tact-lang/emulator";
import { TactWallet, SendParameters, Array as ton_Array } from "./output/sample_TactWallet";
import { fill_send_parameters, send_ext_message } from "./helpers";
import { inspect } from "util";
import { Dictionary } from "ton-core";

describe("contract", () => {
    it("should deploy correctly", async () => {
        let system = await ContractSystem.create();
        let creator = system.treasure("creator");
        let mnemonics = await mnemonicNew();
        let pair = await mnemonicToWalletKey(mnemonics);
        const map = Dictionary.empty<any, any>();
        map.set(0, BigInt("0x" + pair.publicKey.toString("hex")));
        let wallet = system.open(await TactWallet.fromInit({ $$type: "Array", length: 1n, map } as ton_Array));
        await wallet.send(creator, { value: toNano(10) }, "Hello");
        await system.run();
        expect(await wallet.getGetPublicKeys().then((a) => a["_map"])).toEqual(
            ({ $$type: "Array", length: 1n, map } as ton_Array).map["_map"]
        );
        expect(await wallet.getSeqno()).toEqual(0n);

        let params: SendParameters = fill_send_parameters(
            creator.address,
            toNano("0.1"),
            beginCell().storeUint(0, 32).storeStringTail("Hello Creator!").endCell()
        );
        let tracker = system.track(wallet.address);
        await send_ext_message(wallet, pair.secretKey, BigInt(system.now + 15), params);
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
        expect(await wallet.getSeqno()).toEqual(1n);
    });

    it("should work for both addresses", async () => {
        let system = await ContractSystem.create();
        let creator = system.treasure("creator");
        let mnemonics = await mnemonicNew();
        let pair1 = await mnemonicToWalletKey(mnemonics);
        let pair2 = await mnemonicToWalletKey(mnemonics);
        const map = Dictionary.empty<any, any>();
        map.set(0, BigInt("0x" + pair1.publicKey.toString("hex")));
        map.set(0, BigInt("0x" + pair2.publicKey.toString("hex")));
        let wallet = system.open(await TactWallet.fromInit({ $$type: "Array", length: 1n, map } as ton_Array));
        await wallet.send(creator, { value: toNano(10) }, "Hello");
        await system.run();
        expect(await wallet.getGetPublicKeys().then((a) => a["_map"])).toEqual(
            ({ $$type: "Array", length: 1n, map } as ton_Array).map["_map"]
        );
        expect(await wallet.getSeqno()).toEqual(0n);

        let params: SendParameters = fill_send_parameters(
            creator.address,
            toNano("0.1"),
            beginCell().storeUint(0, 32).storeStringTail("Hello Creator!").endCell()
        );
        let tracker = system.track(wallet.address);
        await send_ext_message(wallet, pair2.secretKey, BigInt(system.now + 15), params);
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
        expect(await wallet.getSeqno()).toEqual(1n);
    });
});
