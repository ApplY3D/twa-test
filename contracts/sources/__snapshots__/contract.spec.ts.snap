// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract should deploy correctly 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000007",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C1252ABEA9AAEFC4991BF19FDC5195F4816F8B7778CEECDF88255D03ECE2086EC6F08EA264C13CA644F4B94D838A27537A9227F5B1DCBFF723F4EECE502A3A4EB672A60B000000006660AD46}
 x{C002483CBCA53D072C89FB2BD8C2CC471AB55F00936FD9D7C95DA2471BB40C308640000000000000000000000000000000000000000000000000000000002FAF080000000000000000000000000000000000000000000000000000000000000000064_}
  x{0000000048656C6C6F2043726561746F7221}",
            "type": "cell",
          },
          "to": "kQAOHdpm6as-dS__crnYXwEmdeMoWyyP9sTqnz9xJ2Ol7CVN",
          "type": "external-in",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11084n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello Creator!",
              "type": "text",
            },
            "bounce": true,
            "from": "kQAOHdpm6as-dS__crnYXwEmdeMoWyyP9sTqnz9xJ2Ol7CVN",
            "to": "@treasure(creator)",
            "type": "internal",
            "value": "0.1",
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should deploy correctly 2`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000007",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C1252ABE1A57B02ECA3B6312E38BC406469F2EA3876FBBC0DB8304D768F4068AB0623E559F0347C88E10C7B63770DBEDB7CC700EB83F47E2738450A2C3980E0CD406A009000000006660AD47}
 x{C002483CBCA53D072C89FB2BD8C2CC471AB55F00936FD9D7C95DA2471BB40C308640000000000000000000000000000000000000000000000000000000002FAF080000000000000000000000000000000000000000000000000000000000000000064_}
  x{0000000048656C6C6F2043726561746F7221}",
            "type": "cell",
          },
          "to": "kQClyqc82bVvuchsinNlgw1XCNfskqRrZksVYe2ViXcoT02O",
          "type": "external-in",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11084n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello Creator!",
              "type": "text",
            },
            "bounce": true,
            "from": "kQClyqc82bVvuchsinNlgw1XCNfskqRrZksVYe2ViXcoT02O",
            "to": "@treasure(creator)",
            "type": "internal",
            "value": "0.1",
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should work for both 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000007",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C1252ABEF0486F0DE6C9B680822D9CD2DDAADFF130BBE94B4A2D35889A308F9A91F197BA1710EA875C9D426AC451EB7656AA97E6847553183DACE2FB87DF4E52740A850D000000006660AD5A}
 x{C002483CBCA53D072C89FB2BD8C2CC471AB55F00936FD9D7C95DA2471BB40C308640000000000000000000000000000000000000000000000000000000002FAF080000000000000000000000000000000000000000000000000000000000000000064_}
  x{0000000048656C6C6F2043726561746F7221}",
            "type": "cell",
          },
          "to": "kQCrSAId2aeTq3lGjXLuECibp2CbcX6BTAgquxlFWYGPaDvK",
          "type": "external-in",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11084n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello Creator!",
              "type": "text",
            },
            "bounce": true,
            "from": "kQCrSAId2aeTq3lGjXLuECibp2CbcX6BTAgquxlFWYGPaDvK",
            "to": "@treasure(creator)",
            "type": "internal",
            "value": "0.1",
          },
        ],
      },
    ],
  },
]
`;

exports[`contract should work for both addresses 1`] = `
[
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000007",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{C1252ABE8F46895F115777F050122D262FA75B01EE120D2DFA9D63AFE6DB803C4C047556A1B38E0C4751357BA8577322AEA476A0C4898725A6E113F9159CC830BF2AE20A0000000066613CCB}
 x{C002483CBCA53D072C89FB2BD8C2CC471AB55F00936FD9D7C95DA2471BB40C308640000000000000000000000000000000000000000000000000000000002FAF080000000000000000000000000000000000000000000000000000000000000000064_}
  x{0000000048656C6C6F2043726561746F7221}",
            "type": "cell",
          },
          "to": "kQBNqiiP3k5fDf6Q0Psee8xL1PayaMwe_EicwsY6ki1xlxIY",
          "type": "external-in",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11084n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello Creator!",
              "type": "text",
            },
            "bounce": true,
            "from": "kQBNqiiP3k5fDf6Q0Psee8xL1PayaMwe_EicwsY6ki1xlxIY",
            "to": "@treasure(creator)",
            "type": "internal",
            "value": "0.1",
          },
        ],
      },
    ],
  },
]
`;
